art/defences/trainer/adversarial_trainer_oaat_pytorch.py

```diff
@@ -0,0 +1,1224 @@
+                    i_epoch,
+                    train_time - start_time,
+                    train_loss / train_n,
+                    train_acc / train_n,
+                    nb_correct_clean / x_test.shape[0],
+                    nb_correct_adv / x_test.shape[0],
+                )
+
+                # save last checkpoint
+                if i_epoch + 1 == nb_epochs:
+                    self._classifier.save(filename=f"oaat_epoch_{i_epoch}", path=self._train_params["models_path"])
+
+                # save best checkpoint
+                if nb_correct_adv / x_test.shape[0] > best_acc_adv_test:
+                    self._classifier.save(filename="oaat_epoch_best", path=self._train_params["models_path"])
+                    best_acc_adv_test = nb_correct_adv / x_test.shape[0]
+
+            else:
+                logger.info(
+                    "epoch: %s time(s): %.1f loss: %.4f acc-adv: %.4f",
+                    i_epoch,
+                    train_time - start_time,
+                    train_loss / train_n,
+                    train_acc / train_n,
+                )
+
+            if i_epoch > self._train_params["swa_save_epoch"] - 1:
+                for epoch_swa, tau_val, p_classifier in zip(
+                    self._train_params["list_swa_epoch"], self._train_params["list_swa_tau"], self._list_avg_models
+                ):
+                    if epoch_swa <= i_epoch:
+                        file_name = f"oaat_swa_tau_{tau_val}.model"
+                        if self._train_params["models_path"] is None:
+                            full_path = os.path.join(config.ART_DATA_PATH, file_name)
+                        else:
+                            full_path = os.path.join(self._train_params["models_path"], file_name)
+
+                        folder = os.path.split(full_path)[0]
+                        if not os.path.exists(folder):
+                            os.makedirs(folder)
+                        # pylint: disable=W0212
+                        # disable pylint because access to _modules required
+                        torch.save(p_classifier._model._model.state_dict(), full_path)
+
+    def fit_generator(
+        self,
+        generator: DataGenerator,
+        validation_data: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+        nb_epochs: int = 20,
+        **kwargs,
+    ):  # pylint: disable=W0221
+        """
+        Train a model adversarially with OAAT protocol using a data generator.
+        See class documentation for more information on the exact procedure.
+
+        :param generator: Data generator.
+        :param validation_data: Tuple consisting of validation data, (x_val, y_val)
+        :param nb_epochs: Number of epochs to use for trainings.
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function of
+                                  the target classifier.
+        """
+        import os
```

Comment:
Please move import of `os` to top-level.

Reply:
moved.